### DISCUSSION TITLE
Work around for pydantic_core._pydantic_core.Url in V2 where string is expected

### DISCUSSION THREAD

**Discussion by iamchrisearle** [pydantic/pydantic#8211] [Question]
As mentioned [here](https://github.com/pydantic/pydantic/issues/7186#issuecomment-1691594032)
> In pydantic v1 urls inherited from str, so it can be easily passed to library codes, where they usually expects string representation of urls.

For example:
```python
import os
from urllib.parse import urlsplit

from pydantic import AnyUrl as AnyUrlv2
from pydantic import BaseModel, version
from pydantic.v1 import AnyUrl as AnyUrlv1

version.version_info()

any_url_v1 = AnyUrlv1("http://www.example.com", scheme="http", host="www.example.com")
any_url_v2 = AnyUrlv2("http://www.example.com")

type(any_url_v1)
# returns: <class 'pydantic.v1.networks.AnyUrl'>

type(any_url_v2)
# returns: <class 'pydantic_core._pydantic_core.Url'>

urlsplit("https://google.com/")
os.fspath("https://google.com/")
# both function fine, as these are strings.

urlsplit(any_url_v1)
os.fspath(any_url_v1)
# both function fine, as these inherit from strings.

urlsplit(any_url_v2)
# Errors with AttributeError: 'pydantic_core._pydantic_core.Url' object has no attribute 'decode'
os.fspath(any_url_v2.url)
# Errors with TypeError: expected str, bytes or os.PathLike object, not Url
```

V1 contained very useful functionality by being string like but also containing accessible attributes such as `.host`. This functionality is lost when converting AnyUrls in V2 to strings first before passing to some of the example issues above. Does anyone have a work around for this that doesn't requires coercing to string first?

Version info:
```
            pydantic version: 2.5.1
       pydantic-core version: 2.14.3
         pydantic-core build: profile=release pgo=false
                 install path: /Users/chris/opt/miniconda3/envs/test_pydantic_2/lib/python3.9/site-packages/pydantic 
             python version: 3.9.18 | packaged by conda-forge | (main, Aug 30 2023, 03:52:10)  [Clang 15.0.7 ]
                    platform: macOS-13.6-x86_64-i386-64bit
            related packages: typing_extensions-4.8.0'
```

**Comment by damascene** [comment]
I've spent hours trying to solve this and started writing another thread but just to find yours before pressing the start discussion button. Here is my issue:

I'm using Fastapi with Pydentic to validated data and I set url filed to be an `HttpUrl`:

```python
class Data(BaseModel):
    name: str
    url: HttpUrl 
```

then I'm using it to create toml string:

```python

@app.post("/submit/")
async def create_item(submit: Model):
    try:
         usr_data = ur_data.dict()
         toml_data = tomli_w.dumps(usr_data)
```

I'm getting the following error:

```
INFO:root:create_item Object of type <class 'pydantic_core._pydantic_core.Url'> is not TOML serializable
```

I've also tested it with json dump and found similar issue:
```python
json.dumps(ur_data.dict())
"""output:
TypeError: Object of type Url is not JSON serializable
"""
```

It seems because the field url now is showing as a pydentic class instead of URL string. 

```
'url': Url('https://example.com/')
```

Can I just have it stored as a string without the "pydantic_core._pydantic_core.Url" class?

seems that someone else [raised a similar issue](https://stackoverflow.com/questions/76385999/how-to-export-a-pydantic-model-instance-as-yaml-with-url-type-as-string) but solution seems too complicated for me.

**Reply by iamchrisearle** [reply]
I'm not sure of what your specific situation is, but if you can use V1 without breaking dependencies, AnyURL will serialize to toml:

```python
import tomli_w
from pydantic.v1 import AnyUrl, BaseModel


class Data(BaseModel):
    name: str
    url: AnyUrl


usr_data = {"name": "abc", "url": "https://example.com/"}

result = Data.parse_obj(usr_data)

tomli_w.dumps(dict(result))
# >>> tomli_w.dumps(dict(result))
# 'name = "abc"\nurl = "https://example.com/"\n'


usr_data_bad = {"name": "abc", "url": "totally_not_a_url"}

# double check it still validates a URL as expected
result_bad = Data.parse_obj(usr_data_bad)
```

I say this hesitantly as it could introduce deprecation issues with`parse_obj` and `AnyURL`, but perhaps will get you out of a bind...

**Reply by damascene** [reply]
Thank you, Your code is very good as an example to reproduce the issue:

```python
import tomli_w
from pydantic import AnyUrl, BaseModel


class Data(BaseModel):
    name: str
    url: AnyUrl


usr_data = {"name": "abc", "url": "https://example.com/"}

result = Data.parse_obj(usr_data)

print(tomli_w.dumps(dict(result)))

output = """
test_pydentic.py:12: PydanticDeprecatedSince20: The `parse_obj` method is deprecated; use `model_validate` instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.5/migration/
  result = Data.parse_obj(usr_data)
Traceback (most recent call last):
  File "test_pydentic.py", line 14, in <module>
    print(tomli_w.dumps(dict(result)))
  File "/home/user/.cache/pypoetry/virtualenvs/staging-hur-_E5WuZ78-py3.8/lib/python3.8/site-packages/tomli_w/_writer.py", line 39, in dumps
    return "".join(gen_table_chunks(__obj, ctx, name=""))
  File "/home/user/.cache/pypoetry/virtualenvs/staging-hur-_E5WuZ78-py3.8/lib/python3.8/site-packages/tomli_w/_writer.py", line 73, in gen_table_chunks
    yield f"{format_key_part(k)} = {format_literal(v, ctx)}\n"
  File "/home/user/.cache/pypoetry/virtualenvs/staging-hur-_E5WuZ78-py3.8/lib/python3.8/site-packages/tomli_w/_writer.py", line 102, in format_literal
    raise TypeError(f"Object of type {type(obj)} is not TOML serializable")
TypeError: Object of type <class 'pydantic_core._pydantic_core.Url'> is not TOML serializable
"""
```

However I guess the pydentic team may save people time by adding some flag to disable this v2 behavior, right?

**Reply by iamchrisearle** [reply]
> the pydentic team may save people time by adding some flag to disable this v2 behavior, right?

I can't speak to future release behaviour on this.
However, I did notice this [work around](https://github.com/pydantic/pydantic/issues/7186#issuecomment-1874338146) from @pavradev that could help with keeping you in V2 while allowing serialization. I have tested locally for the simple example above. 

Note that `Annotated` is from typing (`from typing import Annotated`)

**Reply by damascene** [reply]
Thank you, I used that but it shows in Swagger that the expected value is string not URL however it shows a proper response when URL is not provided.

**Reply by iamchrisearle** [reply]
If you are limited by a string-like requirement downstream for your URL can I ask what the purpose of the Data() model is in the first place? Would `.contains("http")` or some other [validator](https://docs.pydantic.dev/latest/concepts/validators/) work for your use case? Just a thought. I'm sure others are more familiar with the specific package functionality and used cases that are more tangential to yours.

**Comment by patrsc** [comment]
I created a Python package [pydantic-string-url](https://pypi.org/project/pydantic-string-url/), which can be used as a replacement for Pydantic's URL types `AnyUrl`, `AnyHttpUrl`, `HttpUrl`, `AnyWebsocketUrl`, `WebsocketUrl`, `FileUrl`, and `FtpUrl`. Those types are based on `str` and can be directly passed to library packages that expect strings. They also preserve the absence of trailing slashes. Usage example:

```py
from pydantic import BaseModel, TypeAdapter, ValidationError
from pydantic_string_url import HttpUrl

class User(BaseModel):
    name: str
    age: int
    homepage: HttpUrl

user = {"name": "John Doe", "age": 33, "homepage": "https://example.com"}
invalid_user = {"name": "Alice", "age": 32, "homepage": "not a url"}

john = User.model_validate(user)
assert john.homepage == "https://example.com"  # no trailing slash was added
try:
    alice = User.model_validate(invalid_user)
except ValidationError as e:
    print(str(e))
```

[Feedback](https://github.com/patrsc/pydantic-string-url/issues) and bug reports are welcome. Maybe something similar makes it into Pydantic in the future.
